File: server/siren/siren.go
package siren

import (
	"slices"
)

type Entity struct {
	Actions    []Action
	Entities   []Entity
	Links      []Link
	Properties map[string]interface{}
	Class      []string
	Href       string
}

func (e *Entity) ClassIs(classes ...string) bool {
	return slices.Equal(e.Class, classes)
}

// FindLinkWithRel returns the first link with the given rels.
func (e *Entity) FindLinkWithRel(rels ...string) (*Link, bool) {
	for _, link := range e.Links {
		if slices.Equal(link.Rel, rels) {
			return &link, true
		}
	}
	return nil, false
}

func (e *Entity) StringProperty(key string) (string, bool) {
	if value, ok := e.Properties[key].(string); ok {
		return value, true
	}
	return "", false
}

type Action struct {
	Class                     []string
	Name, Method, Href, Title string
	Typ                       string `json:"type"`
	Fields                    []Field
}

type Field struct {
	Name, Title string
	Typ         string `json:"type"`
	Value       interface{}
}

type Link struct {
	Rel  []string
	Href string
}

File: app/src/lib/politemall/schema/siren.ts
import { z } from "zod";

export type SirenEntity = {
  class: (string | null | undefined)[];
  rel?: string[];
  actions?: SirenAction[];
  entities?: SirenEntity[];
  links?: SirenLink[];
  properties?: Record<string, any>;
  href?: string;
};

export type SirenLink = { class?: string[]; rel: string[]; href: string };
export type SirenAction = {
  href: string;
  method: string;
  name: string;
  fields?: { name: string; type: string; value?: any; title?: string }[];
  class?: string[];
  title?: string;
  type?: string;
};

export const sirenEntity: z.ZodType<SirenEntity> = z.lazy(() =>
  z.object({
    actions: z
      .object({
        href: z.string(),
        method: z.string(),
        name: z.string(),
        fields: z
          .object({ name: z.string(), type: z.string(), value: z.any(), title: z.string().optional() })
          .array()
          .optional(),
        class: z.string().array().optional(),
        title: z.string().optional(),
        type: z.string().optional(),
      })
      .array()
      .optional(),
    entities: sirenEntity.array().optional(),
    links: z
      .object({ class: z.string().array().optional(), rel: z.string().array(), href: z.string() })
      .array()
      .optional(),
    rel: z.string().array().optional(),
    properties: z.record(z.string(), z.any()).optional(),
    class: z.string().nullable().array(),
    href: z.string().optional(),
  })
);
